src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java

```diff
@@ -0,0 +1,235 @@
+    private void onRequestFailed(
+            Request<?> request,
+            OnRequestComplete callback,
+            IOException exception,
+            long requestStartMs,
+            @Nullable HttpResponse httpResponse,
+            @Nullable byte[] responseContents) {
+        try {
+            NetworkUtility.handleException(
+                    request,
+                    callback,
+                    exception,
+                    requestStartMs,
+                    httpResponse,
+                    responseContents,
+                    this);
+        } catch (VolleyError volleyError) {
+            callback.onError(volleyError);
+        }
+    }
+
+    @Override
+    public void performRequest(final Request<?> request, final OnRequestComplete callback) {
+        mBlockingExecutor = getBlockingExecutor();
+        if (mBlockingExecutor == null) {
+            throw new IllegalStateException(
+                    "mBlockingExecuter should be set before making a request");
+        }
+        final long requestStartMs = SystemClock.elapsedRealtime();
+        // Gather headers.
+        final Map<String, String> additionalRequestHeaders =
+                HttpHeaderParser.getCacheHeaders(request.getCacheEntry());
+        mAsyncStack.executeRequest(
+                request,
+                additionalRequestHeaders,
+                new AsyncHttpStack.OnRequestComplete() {
+                    @Override
+                    public void onSuccess(HttpResponse httpResponse) {
+                        onRequestSucceeded(request, requestStartMs, httpResponse, callback);
+                    }
+
+                    @Override
+                    public void onAuthError(AuthFailureError authFailureError) {
+                        callback.onError(authFailureError);
+                    }
+
+                    @Override
+                    public void onError(IOException ioException) {
+                        onRequestFailed(
+                                request,
+                                callback,
+                                ioException,
+                                requestStartMs,
+                                /* httpResponse= */ null,
+                                /* responseContents= */ null);
+                    }
+                });
+    }
+
+    /* Helper method that determines what to do after byte[] is received */
+    private void onResponseRead(
+            long requestStartMs,
+            int statusCode,
+            HttpResponse httpResponse,
+            Request<?> request,
+            OnRequestComplete callback,
+            List<Header> responseHeaders,
+            byte[] responseContents) {
+        // if the request is slow, log it.
+        long requestLifetime = SystemClock.elapsedRealtime() - requestStartMs;
+        logSlowRequests(requestLifetime, request, responseContents, statusCode);
+
+        if (statusCode < 200 || statusCode > 299) {
+            onRequestFailed(
```

Comment:
missing short-circuit return. Is it possible to unit test for this and/or do a deeper look for cases where this is happening?

Reply:
Are you asking whether it's possible to unit test whether we return correctly? The only thing I can think of is verifying that methods returning are never called. But that relies on us knowing where they all are. I think just making different types of requests in a test app and making sure they don't make two callback calls is the best way to do it.

Reply:
Right, my hope is that it would be possible to wire up a request with a mock stack and validate that the expected callback is invoked exactly once and that the other callback methods are not invoked at all.

Reply:
okay, I'll work on that. I think that shouldn't be too hard actually.